### PR TITLE
catalog-backend: fix handling of optional locations

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

This fixes the catalog warning log spam that you would get from registering something like `https://github.com/backstage/*/blob/-/catalog-info.yaml` as a `github-discovery` location in the catalog. The intention was that each `catalog-info.yaml` location that the discoverer produced would be optional, but this wasn't connected all the way in the catalog processing and so warnings were being logged anyway.

This also required the addition of the `spec.presence` field for Location entities, which was copied over from the `LocationSpec`.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/catalog-backend/src/processing/ProcessorOutputCollector.ts [hunk 1]
```diff
@@ -105,8 +105,9 @@ export class ProcessorOutputCollector {
 
       this.deferredEntities.push({ entity, locationKey: location });
     } else if (i.type === 'location') {
+      const presence = i.optional ? 'optional' : 'required';
       const entity = locationSpecToLocationEntity(
-        i.location,
+        { presence, ...i.location },
```
Comment (Rugvip) [inline plugins/catalog-backend/src/processing/ProcessorOutputCollector.ts:110]: Went for backwards compatibility for now, we'd ofc get rid of the `optional` parameter at some point as we look at the processing API again

File: packages/catalog-model/api-report.md [hunk 2]
```diff
@@ -325,6 +325,7 @@ interface LocationEntityV1alpha1 extends Entity {
     type?: string;
     target?: string;
     targets?: string[];
+    presence?: 'required' | 'optional';
```
Comment (Rugvip) [inline packages/catalog-model/api-report.md:328]: Let's not merge this without debating this addition for a bit :grin:

**Comment by benjdlambert** [comment]
If I'm not mistaken we had support for optional locations before the big re-write. I think we might have skipped this when building `v2` right?

**Review by freben — APPROVED** [review]

File: packages/catalog-model/api-report.md (hunk 2 above)
Comment (Rugvip) [reply packages/catalog-model/api-report.md:328]: Debate was had

**Review by benjdlambert — APPROVED** [review]

**Pull request merged**